src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationHandler.cs

```diff
@@ -99,5 +107,69 @@ protected override Task<bool> HandleForbiddenAsync(ChallengeContext context)
         {
             throw new NotSupportedException();
         }
+
+        protected virtual void GenerateCorrelationId(AuthenticationProperties properties)
+        {
+            if (properties == null)
+            {
+                throw new ArgumentNullException(nameof(properties));
+            }
+
+            var nonceBytes = new byte[32];
+            CryptoRandom.GetBytes(nonceBytes);
+            var correlationId = Base64UrlTextEncoder.Encode(nonceBytes);
+
+            var cookieOptions = new CookieOptions
+            {
+                HttpOnly = true,
+                Secure = Request.IsHttps,
+                Expires = properties.ExpiresUtc
+            };
+
+            properties.Items[CorrelationPrefix] = correlationId;
+
+            Response.Cookies.Append(CorrelationPrefix + correlationId, CorrelationMarker, cookieOptions);
+        }
+
+        protected virtual bool ValidateCorrelationId(AuthenticationProperties properties)
+        {
+            if (properties == null)
+            {
+                throw new ArgumentNullException(nameof(properties));
+            }
+
+            string correlationId;
+            if (!properties.Items.TryGetValue(CorrelationPrefix, out correlationId))
+            {
+                Logger.LogWarning(26, "{0} state property not found.", CorrelationPrefix);
+                return false;
+            }
+
+            properties.Items.Remove(CorrelationPrefix);
+
+            var cookieName = CorrelationPrefix + correlationId;
+
+            var correlationCookie = Request.Cookies[cookieName];
+            if (string.IsNullOrEmpty(correlationCookie))
+            {
+                Logger.LogWarning(27, "{0} cookie not found.", cookieName);
+                return false;
+            }
+
+            var cookieOptions = new CookieOptions
+            {
+                HttpOnly = true,
+                Secure = Request.IsHttps
+            };
+            Response.Cookies.Delete(cookieName, cookieOptions);
+
+            if (!string.Equals(correlationCookie, CorrelationMarker, StringComparison.Ordinal))
+            {
+                Logger.LogWarning(28, "{0} correlation cookie and state property mismatch.", cookieName);
```

Comment:
Options:
* Change the way we generate the cookie to store the generated bits in the cookie value instead of the name (or a few bits in the name and the rest in the value).
* Remove the check, as it's basically useless.
* Move the log message to the appropriate place (i.e L. 158)

@Tratcher @Eilon any preference?